[api] Filter encrypted attributes in responses

abellotti opened the pull request:
- Added encrypted_columns to the password mixin models allowing one to query things like MiqDatabase.encrypted_columns
- Added a helper method to return all encrypted attributes via ApiController.all_encrypted_attributes
- Updated api normalizer to support encrypted attribute type
- Filter encrypted attributes as [FILTERED]

abellotti commented:
@Fryguy as discussed. right now the classes of interest are hard coded in Authentication.all_encrypted_columns, I wish there was a way to dynamically reference the classes that include the password mixin.  Also, in rest api, I return encrypted attributes as "[FILTERED]", optionally I can exclude them from the responses. Let me know.  Thanks.

Fryguy commented:
I don't understand the need for having the all_encrypted columns method.  I would expect that when you get to the piece that grab attributes for a specific class, you would do somehting like `auth.attributes.except(auth.class.encrypted_columns)`, instead of a global lookup you have to maintain when someone adds a new encrypted column.

Fryguy commented on vmdb/app/models/mixins/password_mixin.rb:
```diff
@@ -6,6 +6,15 @@ module PasswordMixin
+
+    def add_encrypted_column(column)
+      @encrypted_columns ||= []
+      @encrypted_columns |= [column] if column.present?
```
just do `<<` ?

Also, thinking this should be a class variable `@@`

Fryguy replied:
Actually, you don't need this method at all.

Down below on line 25, just do

```ruby
encrypted_columns << column.to_s
encrypted_columns << "#{column}_encrypted"

```

abellotti replied:
reason I went with |= vs <<, I wanted uniqueness.

Fryguy replied:
Seems unnecessary...I doubt we'd ever double-call encrypt_column

abellotti commented:
2 reasons for that, one, somehow the AuthUseridPassword.encrypted_columns returns [] even though it exposes password via subclassing, also, with the all_encrypted_columns approach, the Rest API code change was cleaner and allows for filtering vs. hiding.

Fryguy commented:
>> somehow the AuthUseridPassword.encrypted_columns returns [] even though it exposes password via subclassing,

Because you are using class instance variables.  You need to either use `@@` class variables, or use class_attribute (from Rails)

abellotti commented:
yep, works better with class variable. updated accordingly.  Thanks.

Fryguy commented on vmdb/app/controllers/api_controller.rb:
```diff
@@ -95,6 +96,13 @@ def initialize
+  def self.all_encrypted_attributes
+    (ATTR_TYPES[:encrypted]             |
+     ::Authentication.encrypted_columns |
+     ::MiqDatabase.encrypted_columns    |
```
I really don't like harboring these classes... Just feels wrong.

abellotti replied:
I do need my list for the normalizer, considering I can be rendering regular hashes like configurations, provisioning requests options, etc and not just objects.  One option is to update my list dynamically upon rendering db objects.  Let me think about it further.

abellotti replied:
Now dynamically growing that list upon normalizing objects. Thanks.

Fryguy replied:
LOL I just noticed it said harboring.  Silly autocorrect.  I really don't like *hardcoding* these classes.

Fryguy commented on vmdb/app/models/mixins/password_mixin.rb:
```diff
@@ -13,6 +14,9 @@ def encrypt_column(column)
       #   password_encrypted  : Get the password in cryptext
       #   password_encrypted= : Set the password in cryptext
 
+      self.encrypted_columns ||= []
```
This should not be necessary... The method itself initializes.

Fryguy replied:
Oh no wait you removed that... I like it better the other way where it initialized itself on first call

abellotti commented:
@Fryguy changes made as you mentioned, still safe as we dynamically get our list upon rendering. repushed PR. Thanks.

kbrock commented on vmdb/app/controllers/api_controller.rb:
```diff
@@ -95,6 +96,10 @@ def initialize
   end
 
+  def self.all_encrypted_attributes
+    (ATTR_TYPES[:encrypted] | ::Vmdb::ConfigurationEncoder::PASSWORD_FIELDS).sort
```
yay.

As noted before, there are a few other password fields. but I don't think any of those values should come back

Fryguy replied:
Why create a method when this can just live in ATTR_TYPES directly.  On that note, why is ConfigurationEncoder involved at all?  Those PASSWORD_FIELDS are embedded into a configuration yaml, which I doubt you're ever going to return over the REST API.  None of them are columns, so it's not necessary.

kbrock replied:
these are embedded in the yaml returning from the api.
so while it is not coming back as a column, it is a key within settings in those containers.
And those yaml hashes are coming over the wire in things like `provisioning_requests`

Fryguy replied:
So, are we returning the config itself or a copy of it?  Why would the REST API return it at all?

abellotti replied:
They are accessible via provisioning request options hash and configurations via virtual attribute of servers. We just need to make sure we cover all cases.  Thanks.

kbrock commented:
Where is this? :+1: LGTM

kbrock commented on vmdb/app/models/mixins/password_mixin.rb:
```diff
@@ -13,6 +18,9 @@ def encrypt_column(column)
       #   password_encrypted= : Set the password in cryptext
 
+      encrypted_columns << column.to_s
+      encrypted_columns << "#{column}_encrypted"
```
@gmcculloug in automate, do we do something similar to this?
(not trying to de-dup the code, just want it to behave similarly in both spots.)

kbrock replied:
not sure that `column_encrypted` is an `encrypted_columns`. Thought only `column.to_s` would be in there

abellotti replied:
if it's a virtual attribute, it's accessible via the API.

Fryguy replied:
> not sure that column_encrypted is an encrypted_columns. Thought only column.to_s would be in there

Interesting point, and maybe that means the method name is bad.  I too would expect that a method named `encrypted_columns` would not contain `column_encrypted`.  The use case for this method is really being able to filter specific values from the REST API, and we definitely want to filter the column_encrypted.

This is tricky...maybe it just needs a better name, or perhaps you need two separate methods.

abellotti replied:
Discussed with Jason today, will keep name as is for now.

kbrock commented:
update:

It looks like all passwords are sanitized.
For all field names and for hashes as well. Tables with hashes are accessible: `MiqRequest`, `MiqRequestTask`, and `Configuration`.

`Configuration` was double checked that it was properly sanitizing it. Others should work too.

Alberto and I went through all password special cases found in `FixAuth` and ensured they are checked here. That list is pretty solid and should cover everything.

I'm just waiting to hear from gm to see how we handle `#{field}_encrypted` in automate but after that I'll merge

Fryguy commented on vmdb/app/models/mixins/password_mixin.rb:
```diff
@@ -2,10 +2,15 @@ module PasswordMixin
   extend ActiveSupport::Concern
 
   included do
+    class_attribute :encrypted_column_list
```
I still feel like this is weird having one class_attribute for the list, and then a separate method for the list again, `encrypted_columns` below.  There's gotta be a way to override it properly.

Fryguy commented on vmdb/app/controllers/api_controller/initializer.rb:
```diff
@@ -95,23 +106,34 @@ def new_token_mgr(mod, name, api_config)
+          next if cspec[:klass].blank?
+          klass = cspec[:klass].constantize
+          klass.columns_hash.collect  do |name, typeobj|
+            @attr_time[name] = true if %w(date datetime).include?(typeobj.type.to_s)
```
So, I'm confused by the entire design in this PR.  You have a method here that loops through "the Classes we care about", and sets time metadata when it encounters a date column.  Why can't this just be extended in the same way for encrypted attributes?

```ruby
@attr_encrypted[name] = true if klass.encrypted_columns.include?(name)
```

Then you don't need the global "all attributes" method.  I must be missing something, because this seems way over-complicated.

abellotti replied:
I initially wanted the all encrypted attrs approach, thus the method. Then we went with dynamic fetching this stuff. The default list that I need (default, config, prov hashes) can move to the ATTR_TYPES list. Thanks

Fryguy commented on vmdb/app/controllers/api_controller/initializer.rb:
```diff
@@ -36,6 +36,17 @@ def collection_config
+        return unless @encrypted_objects_checked[klass].nil?
+        @encrypted_objects_checked[klass] = object_encrypted_attributes(obj)
+        @encrypted_objects_checked[klass].each { |attr| @attr_encrypted[attr] = true }
+      end
```
Why is this method needed...it seems to check for encrypted columns and then caches it, but you already cache that in `@attr_encrypted` on line 118.

abellotti replied:
When fetching tons of vms in a request, I only want to do this once. Plus the method is used for both direct and virtual attribute querying paths. Thanks.

Fryguy commented on vmdb/app/controllers/api_controller.rb:
```diff
@@ -65,8 +65,9 @@ class UnsupportedMediaTypeError < StandardError; end
-    :url  => %w(href)
+    :time      => %w(expires_on),
+    :url       => %w(href),
+    :encrypted => %w(password root_password sysprep_password sysprep_domain_password)
```
And for this, I feel we should leverage the encrypted_columns method on the class in question.  Doesn't make sense to hardcode here something that we will have to remember to keep up to date.

kbrock replied:
hashes keys are not defined via `encrypted column`

Fryguy replied:
But, this is conflating the two things, and is not properly maintainable.

Fryguy commented on vmdb/app/helpers/api_helper/normalizer.rb:
```diff
@@ -91,6 +95,13 @@ def normalize_href(type, value)
+    #
+    def normalize_encrypted(_type, _value)
+      "[FILTERED]"
+    end
```
Should we filter the value or just completely remove the key from the output.  I was originally thinking the latter.  Why even expose that we *have* a password column.

kbrock replied:
:+1:

abellotti replied:
I'm game either way.

Fryguy commented:
@kbrock Don't merge yet until I :+1:

kbrock commented:
:+1: LGTM

kbrock commented on vmdb/app/controllers/api_controller.rb:
```diff
@@ -65,8 +65,10 @@ class UnsupportedMediaTypeError < StandardError; end
+    :time      => %w(expires_on),
+    :url       => %w(href),
+    :encrypted => %w(password root_password sysprep_password sysprep_domain_password) |
+                  ::Vmdb::ConfigurationEncoder::PASSWORD_FIELDS
```
this is simpler. thanks

abellotti commented:
@gmcculloug updated PR as discussed.  Now providing MiqProvisionWorkflow.all_encrypted_options_fields and REST API leveraging this instead of hard coding root_password, sysprep_password, sysprep_domain_password.  Let me know if the changes are ok or they need to be updated as per the upcoming changes in your PRs.  Thanks.

kbrock commented on vmdb/app/controllers/api_controller.rb:
```diff
@@ -66,8 +66,10 @@ class UnsupportedMediaTypeError < StandardError; end
-    :url  => %w(href)
+    :time      => %w(expires_on),
+    :url       => %w(href),
+    :encrypted => %w(password) | ::MiqProvisionWorkflow.all_encrypted_options_fields.map(&:to_s) |
```
of note, these are hardcoded in `FixAuth` logic.
Since it is standalone, can't really do anything about that.

abellotti commented:
@Fryguy @gmcculloug updated as discussed. tested & repushed.  Thanks.

gmcculloug commented:
Changes look good.  @Fryguy Please review.

abellotti commented:
Added Rspecs for this which also unveiled a Jbuilder limitation not honoring ignore_nil in embedded hashes, fixed here in normalizer code (good old rspecs).
